src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -3310,70 +3335,350 @@ new_readdir(DIR *dirp)
- *{
- *	printf("DBG> execve(%s)\n", filename);
- *	return real_execve(filename, argv, pre_envp(envp));
- *}
- *
- *static int
- *new_execvp(const char *filename, char *const argv[])
- *{
- *	printf("DBG> execvp(%s)\n", filename);
- *	return real_execve(filename, argv, pre_envp(NULL));
- *}
- *
- *static int
- *new_execv(const char *filename, char *const argv[])
- *{
- *	printf("DBG> execv(%s)\n", filename);
- *	pre_envp(NULL);
- *	return real_execve(filename, argv, pre_envp(NULL));
- *}
- *
- *static pid_t
- *new_fork(void)
- *{
- *	pid_t pid;
- *	pid = real_fork();
- *	if (pid) {
- *		// parent process: do nothing
- *		return pid;
- *	} else {
- *		init_dfs();
- *		return pid;
- *	}
- *}
+/* This is the number of environmental variables that would be forced to set in child process.
+ * "LD_PRELOAD" is a special case and it is not included in the list.
+ */
+static char  *env_list[] = {"D_IL_REPORT", "DAOS_MOUNT_POINT", "DAOS_POOL", "DAOS_CONTAINER",
+			    "D_IL_MAX_EQ", "D_LOG_FILE", "D_IL_ENFORCE_EXEC_ENV",  "DD_MASK",
+			    "DD_SUBSYS", "D_LOG_MASK"};
+
+/* Environmental variables could be cleared in some applications. To make sure all libpil4dfs
+ * related env properly set, we intercept execve and its variants to check envp[] and append our
+ * env when a stripped env list is provided.
+ * pre_envp() scans provided envp to look for libpil4dfs related envs (env_list[]) and returns a
+ * new envp with missing libpil4dfs related envs appended. Non-zero *error returns errno. 
+ * envp could be NULL, so the return value could be NULL too.
+ * The fd for daos logging may be not valid here, so fprintf() is used in this function.
  */
+static char **
+pre_envp(char *const envp[], int *error)
+{
+	int    i, j, rc;
+	int    len, len2, len_total;
+	int    num_env_append;
+	char **new_envp;
+	char  *pil4df_path;
+	char  *new_preload_str;
+	int    num_entry       = 0;
+	int    num_entry_found = 0;
+	int    idx_preload     = -1;
+	/* whether pil4dfs set in LD_PRELOAD in environ */
+	bool   pil4dfs_set_preload                = false;
+	bool   preload_included                   = false;
+	/* whether pil4dfs in LD_PRELOAD in envp[] */
+	bool   pil4dfs_in_preload                 = false;
+	/* whether env_list entry exists in envp[] */
+	bool   env_found[ARRAY_SIZE(env_list)]    = {false};
+	/* whether env_list entry exists in environ */
+	bool   env_set[ARRAY_SIZE(env_list)]      = {false};
+	/* the buffer allocated for the env string to append */
+	char  *env_buf_list[ARRAY_SIZE(env_list)] = {NULL};
+	char  *env_value                          = NULL;
+
+	/* 0 means no error. */
+	*error = 0;
+	/* simply return for environ. append pil4dfs env only for stripped env list. */
+	if (envp == environ)
+		return (char **)envp;
+
+	/* check D_IL_ENFORCE_EXEC_ENV env in case it was updated. */
+	rc = d_getenv_bool("D_IL_ENFORCE_EXEC_ENV", &enforce_exec_env);
+	if (rc == -DER_NONEXIST || enforce_exec_env == false)
+		return (char **)envp;
```

Comment:
If this environment has been cleaned then this won't be set so the code would never be invoked, at least for the exec case.

You also don't want to read the same environment variable more than once so can you read this at the time daos_init() is called or elsewhere where the library is setup and save the result for use here.

Reply:
Flag enforce_exec_env is set in the constructor function by getenv. We do not monitor whether this env is updated or not. In the case of running a bash script, env "D_IL_ENFORCE_EXEC_ENV" could be updated inside the script. Checking it again will make sure we do get an updated value.